Use OTEL helpers for TLS config instead of tlscfg in es-rollover

chahatsagarmain opened the pull request:
## Which problem is this PR solving?
- Part of #4316 
- 
## Description of the changes
- 

## How was this change tested?
- 

## Checklist
- [ ] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [ ] I have signed all commits
- [ ] I have added unit tests for the new functionality
- [ ] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `yarn lint` and `yarn test`

chahatsagarmain commented:
@yurishkuro i have made some changes here by replacing `tlscfg `with `configtls `, if its in the right direction i can move on to fixing unit test fails due to `tlscfg.Options` in unit tests .

yurishkuro commented on cmd/es-rollover/app/actions.go:
```diff
@@ -35,30 +37,54 @@ type Action interface {
+	Enabled                bool
+}
+
+func (c *ClientConfig) AddFlags(flags *flag.FlagSet) {
```
this is redundant, we should continue using logic from tlscfg for this (see how it's done in cmd/query)

yurishkuro commented on cmd/es-rollover/main.go:
```diff
@@ -129,7 +128,7 @@ func main() {
 		},
 	}
 
-	addPersistentFlags(v, rootCmd, tlsFlags.AddFlags, app.AddFlags)
```
If you remove them here how are the flags registered?

yurishkuro commented on cmd/es-rollover/app/actions.go:
```diff
@@ -35,30 +36,46 @@ type Action interface {
 	Do() error
 }
 
+type ClientConfig struct {
```
why is this type needed?

yurishkuro commented on cmd/es-rollover/app/actions.go:
```diff
@@ -35,30 +36,46 @@ type Action interface {
 
+type ClientConfig struct {
+	configtls.ClientConfig `mapstructure:",squash"`
+	Enabled                bool
```
never used

yurishkuro commented on cmd/es-rollover/app/actions.go:
```diff
@@ -35,30 +36,46 @@ type Action interface {
+	Args      []string
+	Viper     *viper.Viper
+	Logger    *zap.Logger
+	TLSConfig *ClientConfig
```
We can simplify:
```suggestion
	TLSConfig configtls.ClientConfig
```

yurishkuro commented on cmd/es-rollover/app/flags.go:
```diff
@@ -48,6 +52,10 @@ func AddFlags(flags *flag.FlagSet) {
 	flags.Bool(adaptiveSampling, false, "Enable rollover for adaptive sampling index")
 }
 
+func (ClientConfig) AddFlags(flags *flag.FlagSet) {
+	tlsFlagsCfg.AddFlags(flags)
+}
```
```suggestion
func AddTLSFlags(flags *flag.FlagSet) {
	tlsFlagsCfg.AddFlags(flags)
}
```

yurishkuro commented on cmd/es-rollover/main.go:
```diff
@@ -129,7 +128,7 @@ func main() {
 	}
 
-	addPersistentFlags(v, rootCmd, tlsFlags.AddFlags, app.AddFlags)
+	addPersistentFlags(v, rootCmd, tlsConfig.AddFlags, app.AddFlags)
```
```suggestion
	addPersistentFlags(v, rootCmd, AddTLSFlags, app.AddFlags)
```

yurishkuro commented on cmd/es-rollover/main.go:
```diff
@@ -80,10 +79,10 @@ func main() {
+				Args:      args,
+				Viper:     v,
+				Logger:    logger,
+				TLSConfig: tlsConfig,
```
```suggestion
```

yurishkuro commented on cmd/es-rollover/main.go:
```diff
@@ -30,7 +29,7 @@ func main() {
 	}
 
-	tlsFlags := tlscfg.ClientFlagsConfig{Prefix: "es"}
+	tlsConfig := &app.ClientConfig{}
```
not needed

yurishkuro commented on cmd/es-rollover/main.go:
```diff
@@ -42,10 +41,10 @@ func main() {
+				Args:      args,
+				Viper:     v,
+				Logger:    logger,
+				TLSConfig: tlsConfig,
```
not needed

yurishkuro commented on cmd/es-rollover/main.go:
```diff
@@ -109,10 +108,10 @@ func main() {
+				Args:      args,
+				Viper:     v,
+				Logger:    logger,
+				TLSConfig: tlsConfig,
```
not needed

yurishkuro commented on cmd/es-rollover/app/flags.go:
```diff
@@ -63,3 +71,16 @@ func (c *Config) InitFromViper(v *viper.Viper) {
 	c.AdaptiveSampling = v.GetBool(adaptiveSampling)
 }
+
+func (c *ClientConfig) InitFromViper(v *viper.Viper) error {
```
never called? It should be part of `func (c *Config) InitFromViper`

chahatsagarmain commented:
@yurishkuro I have removed the unnecessary client config struct and added `ClientConfig ` to `Config` but it seems actions_test unit is failing maybe because i am not properly handling flags in `flags.go`

yurishkuro commented on cmd/es-rollover/app/actions_test.go:
```diff
@@ -74,21 +73,20 @@ func TestExecuteAction(t *testing.T) {
 		t.Run(test.name, func(t *testing.T) {
 			v := viper.New()
-			tlsFlags := tlscfg.ClientFlagsConfig{Prefix: "es"}
 			command := cobra.Command{}
 			flags := &flag.FlagSet{}
-			tlsFlags.AddFlags(flags)
+			AddTLSFlags(flags)
+			AddFlags(flags)
 			command.PersistentFlags().AddGoFlagSet(flags)
 			v.BindPFlags(command.PersistentFlags())
```
you should be able to replace this with a single call to Viperize from pkg/config, e.g.
```
v, command := config.Viperize(AddTLSFlags , AddFlags)
```

yurishkuro commented on cmd/es-rollover/app/flags.go:
```diff
@@ -62,4 +72,10 @@ func (c *Config) InitFromViper(v *viper.Viper) {
+	if err != nil {
+		return err
+	}
+    c.TLSConfig = opts.ToOtelClientConfig()
```
`make fmt`

yurishkuro commented on cmd/es-rollover/app/flags.go:
```diff
@@ -48,8 +54,12 @@ func AddFlags(flags *flag.FlagSet) {
 	flags.Bool(adaptiveSampling, false, "Enable rollover for adaptive sampling index")
 }
 
+func AddTLSFlags(flags *flag.FlagSet) {
+	tlsFlagsCfg.AddFlags(flags)
+}
```
```suggestion
	flags.Bool(adaptiveSampling, false, "Enable rollover for adaptive sampling index")
	tlsFlagsCfg.AddFlags(flags)
}
```

yurishkuro replied:
don't need a new function

yurishkuro commented on cmd/es-rollover/main.go:
```diff
@@ -42,10 +39,9 @@ func main() {
+				Logger: logger,
 			}, func(c client.Client, cfg app.Config) app.Action {
 				initCfg.Config = cfg
 				initCfg.InitFromViper(v)
```
or just panic in InitFromViper, this is not a library so OK to panic

mahadzaryab1 commented on cmd/es-rollover/app/actions.go:
```diff
@@ -37,28 +39,38 @@ type Action interface {
 // ActionCreatorFunction type is the function type in charge of create the action to be executed
 type ActionCreatorFunction func(client.Client, Config) Action
 
+func getTLSConfig(tlsConfig configtls.ClientConfig, logger *zap.Logger) (*tls.Config, error) {
```
I don't believe you need this helper. See below for OTEL's implementation of `LoadTLSConfig`
```go
// LoadTLSConfig loads the TLS configuration.
func (c ClientConfig) LoadTLSConfig(_ context.Context) (*tls.Config, error) {
	if c.Insecure && !c.hasCA() {
		return nil, nil
	}

	tlsCfg, err := c.loadTLSConfig()
	if err != nil {
		return nil, fmt.Errorf("failed to load TLS config: %w", err)
	}
	tlsCfg.ServerName = c.ServerName
	tlsCfg.InsecureSkipVerify = c.InsecureSkipVerify
	return tlsCfg, nil
}
As you can see, it mostly performs the same validations that this function is performing so calling it directly should be sufficient. 
```

yurishkuro commented on cmd/es-rollover/app/actions.go:
```diff
@@ -49,16 +49,18 @@ type ActionCreatorFunction func(client.Client, Config) Action
-	if err != nil {
-		return err
+	if err := cfg.InitFromViper(opts.Viper); err != nil {
+		opts.Logger.Error("Failed to initialize config from viper",
+			zap.Error(err),
+		)
```
there is no need (ever) to log errors if you are returning it anyway - somewhere upstream the caller will decide to log it.

yurishkuro commented on cmd/es-rollover/app/init/flags.go:
```diff
@@ -40,6 +40,9 @@ func (*Config) AddFlags(flags *flag.FlagSet) {
 // InitFromViper initializes config from viper.Viper.
 func (c *Config) InitFromViper(v *viper.Viper) {
+	if v == nil {
+		panic("viper instance cannot be nil")
```
what is the point of this check? dereferencing nil pointer will panic anyway

yurishkuro commented on cmd/es-rollover/app/flags.go:
```diff
@@ -46,10 +52,11 @@ func AddFlags(flags *flag.FlagSet) {
 
 // InitFromViper initializes config from viper.Viper.
-func (c *Config) InitFromViper(v *viper.Viper) {
+func (c *Config) InitFromViper(v *viper.Viper) error {
```
```suggestion
func (c *Config) InitFromViper(v *viper.Viper) {
```

yurishkuro commented on cmd/es-rollover/app/flags.go:
```diff
@@ -62,4 +69,11 @@ func (c *Config) InitFromViper(v *viper.Viper) {
 	c.AdaptiveSampling = v.GetBool(adaptiveSampling)
+	opts, err := tlsFlagsCfg.InitFromViper(v)
+	if err != nil {
+		return err
```
```suggestion
		panic(err)
```
